### PR TITLE
add support for checking new issue against issue templates

### LINKED ISSUE #8: Check if the issue body is the same as the issue template
This app looks awesome!

Currently, it only checks if the body is empty:

https://github.com/behaviorbot/request-info/blob/68e08a896e0e838110e45b5263a042d6ff4bf260/index.js#L21

However, some repositories have a `.github/ISSUE_TEMPLATE.md` ([example](https://github.com/desktop/desktop/blob/master/.github/ISSUE_TEMPLATE.md)), which GitHub automatically puts in the issue entry box. Could this app check if the issue contents are the same as the template?

An even more awesome feature would be to check if there are any fields in the template that need to be filled out, then ask the user to fill them out.

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #8

 - [x] first pass at the API work to check templates
 - [x] write some tests to verify the behaviour

### CODE REVIEW

**Comment by shiftkey** [comment]
I think this is ready for a review. I mostly just reused the existing test infrastructure, which was very helpful to ensure I hadn't broken anything!

File: index.js [hunk 1]
```diff
@@ -37,6 +38,10 @@ module.exports = robot => {
         if (!(await PullRequestBodyChecker.isBodyValid(body, config, context))) {
           badBody = true
         }
+      } else if (eventSrc === 'issue') {
+        if (!(await IssueBodyChecker.isBodyValid(body, config, context))) {
```
Comment (hiimbex) [inline index.js]: What do you think about check `config.checkIssueTemplate` here?

File: lib/IssueBodyChecker.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  return false
+}
+
+class IssueBodyChecker {
+  static async isBodyValid (body, config, context) {
+    if (!body) {
+      return false
+    }
+
+    if (config.checkIssueTemplate && (await isBodyEqualToTemplate(body, context))) {
```
Comment (hiimbex) [inline lib/IssueBodyChecker.js]: I'd recommend moving this check `config.checkIssueTemplate` in accordance with my previous comment!

File: lib/IssueBodyChecker.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+function fromBase64 (content) {
+  return Buffer.from(content, 'base64').toString()
+}
+
+async function getIssueTemplateBody (context, path) {
+  try {
+    const singleTemplate = await context.github.repos.getContent({
+      owner: context.payload.repository.owner.login,
+      repo: context.payload.repository.name,
```
Comment (hiimbex) [inline lib/IssueBodyChecker.js]: Can we clean this up using some [probot internal APIs](https://probot.github.io/api/latest/classes/context.html#repo):

```js 
const singleTemplate = await context.github.repos.getContent(context.repo({path}))
```

File: lib/IssueBodyChecker.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  } catch (err) {
+    return null
+  }
+}
+
+async function getIssueTemplatePaths (context) {
+  try {
+    const templatesDirectory = await context.github.repos.getContent({
+      owner: context.payload.repository.owner.login,
+      repo: context.payload.repository.name,
```
Comment (hiimbex) [inline lib/IssueBodyChecker.js]: Same as above

File: lib/IssueBodyChecker.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  return []
+}
+
+async function isBodyEqualToTemplate (body, context) {
+  const issueTemplates = await getIssueTemplates(context)
+
+  for (const template of issueTemplates) {
+    if (body.includes(template)) {
```
Comment (hiimbex) [inline lib/IssueBodyChecker.js]: This is my main concern with this PR. What if the template was:
```
Title:
Description:
Issue:
```

and I opened an issue with:
```
Title: blah blah
Description: blah blah blah
Issue: more blah
```

This check will fail. If I'm following correctly, `body` is the string of the issue body. Which will be something like:

`Title: blah blah\nDescription: blah blah blah\nIssue: more blah` 

which will not ["include"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes):

`Title:\nDescription:\nIssue:`

`includes()` as a string method is very literal and doesn't handle the sort of pattern matching or regex that we're looking for here. I think someone opening an issue with a the _exact_ contents of an issue template (ie not filled in) is equally bad to it being blank.

Comment (j-f1) [reply lib/IssueBodyChecker.js]: The checker at the bottom of the file returns `false` (i.e. invalid) if this function returns `true`. This checks that the OP has actually filled out the template and not just left it blank.

File: index.js (hunk 1 above)
Comment (shiftkey) [reply index.js]: This was me following the pattern that `PullRequestBodyChecker` used. I can move it out, just wanting to confirm I should ignore that.

File: lib/IssueBodyChecker.js (hunk 5 above)
Comment (shiftkey) [reply lib/IssueBodyChecker.js]: The check I'm trying to do here is whether the issue body matches a template defined in the repository. I want to catch and handle those cases that @j-f1 calls out - people who just submit the template as-is, without making any changes to the issue body. I'm not interested in doing more advanced checks (like trying to correlate fields and values) currently.

File: index.js (hunk 1 above)
Comment (hiimbex) [reply index.js]: This is very minor, so I'll defer to you on whether matching the existing code or being slightly more efficient is better. Let me know which you decide @shiftkey and happy to merge this whenever.

Comment (shiftkey) [reply index.js]: How about we leave this as-is, and I'll follow up with a PR to lift both this `config.checkIssueTemplate` and `config.checkPullRequestTemplate` out of their relevant components and into here, so they remain consistent?

Comment (shiftkey) [reply index.js]: @hiimbex friendly bump :wave:

**Comment by hiimbex** [comment]
Sorry for the delay! Thanks again for this @shiftkey 🎉

**Pull request merged**